docs/development/design/flyout-redesign.rst

```diff
@@ -0,0 +1,165 @@
+===============================
+Flyout Redesign Design Document
+===============================
+
+.. warning::
+
+    This document details designs that are a **work in progress**.
+
+
+This document describes the design of a new "flyout API"
+to replace our existing :ref:`"footer_html" <api/v2:Undocumented resources and endpoints>` endpoint in APIv2.
+The Read the Docs theme uses this API to put an updated version selector
+with active versions into the lower left menu.
+On other themes, this appears as a floating menu in the lower right.
+
+.. contents:: Contents
+   :local:
+   :backlinks: none
+   :depth: 1
+
+
+Issues with the existing API
+------------------------------
+
+.. figure:: ../../_static/images/design-docs/flyout/flyout-expanded.png
+    :align: right
+    :figwidth: 300px
+    :target: ../../_static/images/design-docs/flyout/flyout-expanded.png
+
+    Read the Docs with an expanded flyout menu in the lower left corner
+
+* The largest problem with the existing ``/api/v2/footer_html`` endpoint
+  is that it returns a blob of HTML.
+  This limits the ability to use this data in ways other than to generate our exact flyout.
+  For example, Sphinx themes other than our own cannot consume this data easily
+  and there's no real integration point for MkDocs at all.
+* Due to how the URLs in the version selector are generated,
+  this API endpoint can be fairly expensive in the worst case for projects with many versions.
+  As it stands now, this API accounts for approximately 15% of the time taken on our webservers
+  at about 70ms for an average request (P95 ~= 150ms, P99 ~= 235ms).
+* The current API is a combination of data about a project that contains information
+  on the live versions, translations, sub- and super-projects, and links to various other things
+  like downloads, the repository on the VCS, and builds for the project on RTD itself.
+  Some of this data never changes (links to the project and builds on RTD itself)
+  and are effectively static while some of it could change after the documentation is built (translations, active versions).
+
+
+Overlap with existing API endpoints
+-----------------------------------
+
+There is already significant overlap between the APIv2 ``footer_html`` call
+with the existing APIv3 for a project (eg. ``https://readthedocs.org/api/v3/projects/docs/``).
+The project API already returns much of the data we want,
+but some things like other active versions, translations, and downloads would require additional API calls or options.
+These options already partially implemented via the `DRF Rest Fields <https://pypi.org/project/drf-flex-fields/>`_ module
+(eg. ``https://readthedocs.org/api/v3/projects/docs/?expand=active_versions``).
+Currently, there are separate API endpoints for translations and downloads,
+but ideally all the data needed to generate a version selector would be available from a single API.
```

Comment:
It seems that we could implement `?expand=translations,subrpojects` and we would have everything we need with one API call: https://readthedocs.org/api/v3/projects/docs/?expand=active_versions,translations,subprojects

One downside that has using this endpoint is that we will be returning much data than we need and that could slow down the response time.

Another thing to consider is that APIv3 is authenticated-only, and we require anonymous access for the flyout.

Reply:
I'm not sold on the idea of sharing the endpoint for the flyout than a regular endpoint. In the future we may want to _add extra data_ to `/api/v3/projects/` and we may be blocked because it increases the response time by some milliseconds that are not acceptable for the flyout endpoint.

I think that having an independent endpoint with more control over it is a win here and will avoid headaches in the future.